src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/ValidateWorkflowsPass.php

```diff
@@ -57,21 +56,26 @@ public function process(ContainerBuilder $container)
      */
     private function getValidator($tag)
     {
-        if ($tag['type'] === 'state_machine') {
+        if ('state_machine' === $tag['type']) {
             $name = 'state_machine';
-            $class = StateMachineValidator::class;
-        } elseif ($tag['marking_store'] === 'scalar') {
-            $name = 'single_place';
-            $class = SinglePlaceWorkflowValidator::class;
+        } elseif ('single_state' === $tag['marking_store']) {
+            $name = 'workflow_single_state';
         } else {
             $name = 'workflow';
-            $class = WorkflowValidator::class;
         }
 
-        if (empty($this->validators[$name])) {
-            $this->validators[$name] = new $class();
+        if (isset($this->validators[$name])) {
+            return $this->validators[$name];
         }
 
-        return $this->validators[$name];
+        switch ($name) {
```

Comment:
It's a bit cumbersome to have a `if / else` then the same `switch`
And IMHO, we could remove the cache layer to make code simpler. it's a premature optimization that does not bring real perf improvement .